MH-13488 userprovider for the d2l brightspace LMS

kristofkeppens opened the pull request:
A userprovider for the d2l brightspace LMS, similar to the moodle and sakai userprovider. 
It uses the brightspace API to retrieve the courses a user is enrolled in and uses this to create extra roles for the user.

frankCls committed:
userprovider for the d2l brightspace LMS

complete api serialization objects

make userdirectory-brightspace a osgi module

add brightspace api client

moved code to correct package

make checkstyle happy

add documentation

clean up template properties config

clean up brighspace api client

add super admin role

add the published ugent version of the library as dependency

fix typo and add page to index

upgrade version

delete lockfiles

fix accidental remove

frankCls committed:
userprovider for the d2l brightspace LMS

complete api serialization objects

make userdirectory-brightspace a osgi module

add brightspace api client

moved code to correct package

make checkstyle happy

add documentation

clean up template properties config

clean up brighspace api client

add super admin role

add the published ugent version of the library as dependency

fix typo and add page to index

upgrade version

delete lockfiles

fix accidental remove

kristofkeppens committed:
add brightspace documentation to nav index

kristofkeppens committed:
add brightspace documentation to nav index

gregorydlogan commented on modules/userdirectory-brightspace/pom.xml:
```diff
@@ -0,0 +1,153 @@
+    </parent>
+    <properties>
+        <opencast.basedir>${project.basedir}/../..</opencast.basedir>
+        <checkstyle.skip>true</checkstyle.skip>
```
This needs to be removed, and checkstyles needs to pass.

gregorydlogan commented on modules/userdirectory-brightspace/pom.xml:
```diff
@@ -0,0 +1,153 @@
+        <dependency>
+            <groupId>be.ugent.brightspace</groupId>
+            <artifactId>valence-sdk-java</artifactId>
+            <version>1.0.1</version>
```
This version should move up into the main pom.

gregorydlogan commented on modules/userdirectory-brightspace/pom.xml:
```diff
@@ -0,0 +1,153 @@
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-all</artifactId>
+            <version>1.10.19</version>
```
Version up to main pom.

gregorydlogan commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+      throw new ConfigurationException(ORGANIZATION_KEY, "not found");
+    }
+
+    logger.debug("creating new MoodleUserProviderInstance for pid=" + pid);
```
MoodleUserProviderInstance?

lkiesow replied:
```suggestion
    logger.debug("creating new brightspace user provider for pid={}", pid);
```

gregorydlogan commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderInstance.java:
```diff
@@ -0,0 +1,274 @@
+  private User loadUserFromBrightspace(String username) {
+    if (this.cache == null) {
+      throw new IllegalStateException("The Brightspace user detail service has not yet been configured");
+    } else if (!"admin".equals(username) && !"".equals(username) && !"anonymous".equals(username)) {
```
Magic strings should be constants IMO.

lkiesow replied:
But these are configurable in `custom.properties` so we should never hard-code them:

```
org.opencastproject.security.admin.user=admin
```

smarquard replied:
These providers have all inherited this code from the original Sakai provider, so changes like this should really be made across all of them, or we move the change higher up the stack.

The logic for user providers not handling admin and anonymous is it's not necessary to resolve them further (because admin has all permissions and anonymous is, well, anonymous). So that check could go into the UserDirectoryService or whatever's calling these providers.

lkiesow replied:
Then we should fix it in the Sakai user provider as well since that's clearly a bug.

> because admin has all permissions and anonymous is, well, anonymous

The admin user's name in configurable and it could be any other string so there is no guarantee that a user named `admin` has all permissions.

gregorydlogan replied:
Agree with @lkiesow, good catch.  I'm wondering if we should merge this as is, and then fix all the providers with a separate PR after this one and #633 go in.

lkiesow replied:
Let's get it in properly. I'll try to provide a reference implementation. Hopefully tomorrow.

lkiesow replied:
As promised, I created a reference implementation by fixing the Moodle user provider in pull request #902

kristofkeppens replied:
I have implemented it in the same way for this user provider

kristofkeppens committed:
upgrade spring-security and jasig cas library to fix issue with https and reverse proxies

kristofkeppens committed:
upgrade spring-security and jasig cas library to fix issue with https and reverse proxies

gregorydlogan commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderInstance.java:
```diff
@@ -0,0 +1,274 @@
+          return null;
+        }
+      } catch (BrightspaceClientException e) {
+        logger.error("User {} not found in Brightspace system", username);
```
Is this really that the user is not found?  This seems more like a communication error to me.

kristofkeppens replied:
I changed the error to "A Brightspace API error occurred, user {} could not be retrieved". Is this clean enough?

gregorydlogan replied:
Yep, perfect.

gregorydlogan commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderInstance.java:
```diff
@@ -0,0 +1,274 @@
+      JaxbOrganization jaxbOrganization = JaxbOrganization.fromOrganization(organization);
+
+      this.brightspaceWebServiceRequests.incrementAndGet();
+      Thread currentThread = Thread.currentThread();
```
Why are we messing with the classloader?

kristofkeppens replied:
Honestly no idea, the code was developed by an external consultant and it seems that this was copy-pasted from the sakai userprovider.

smarquard replied:
I looked into why that's there in the Sakai provider, and can't find an answer - it's been there for quite some time and I can't find the commit or issue that added it.

gregorydlogan replied:
@smarquard I don't see anything obvious regarding why this is needed.  Can you try yanking the loader at your end and see if it breaks anything.  @flyapen, if you have a system to test this against, can you yank it here too and see if it breaks things?

smarquard replied:
I can try on our test system, but not keen to make a possibly-breaking-change in production right now especially as there's no particular problem that we're trying to solve. I think it may be related to caching or something but it would have been done like that in 1.x days.

smarquard replied:
It turns out the Sakai provider was copied from the Ldap provider, and we find this classloader switching there:

https://github.com/opencast/opencast/blob/ceb51e5a42d21997f478b0d84ef674ab4fc310ee/modules/userdirectory-ldap/src/main/java/org/opencastproject/userdirectory/ldap/LdapUserProviderInstance.java#L320

gregorydlogan replied:
Oh man, old code is old.  Now that I look at how this is being used by the LDAP (and Sakai) code, I'm guessing this switching is more or less required.  The factory which creates this provider has the relevant Karaf service bits and hooks into the security system, I think.  It might be worth testing in a test environment to see if this is still required, if you get a chance.  It's always nice to be able to remove black magic from our code :)

gregorydlogan commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/client/BrightspaceClient.java:
```diff
@@ -0,0 +1,40 @@
+
+public interface BrightspaceClient {
+
+  BrightspaceUser findUser(String username) throws BrightspaceClientException;
```
Ideally these would have javadocs.

gregorydlogan commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/client/api/BrightspaceUser.java:
```diff
@@ -0,0 +1,103 @@
+  private Activation activation;
+  private String displayName;
+
+  @JsonCreator
```
We're using a different library for these kinds of annotations across the rest of the codebase.  Any chance this can be adapted to use that library?  I can see this kind of module not getting security updates since it's using a different library that we might not be checking as closely.

kristofkeppens replied:
I can have a look, could you point me in the right direction to where to find the different library ?

gregorydlogan replied:
Sure thing.  Most of our classes with these annotations are prefixed with "Jaxb".  A simple one would be https://github.com/opencast/opencast/blob/develop/modules/common/src/main/java/org/opencastproject/job/api/JaxbJob.java, and its other Jaxb friends in that package.

kristofkeppens replied:
I had a look to change this myself but not really confident enough that I won't break more than it'll fix. We'll have a look to see if we can get this fixed by a developer.

lkiesow commented on modules/userdirectory-brightspace/pom.xml:
```diff
@@ -0,0 +1,153 @@
+
+    </dependencies>
+    <build>
+        <plugins>
```
Can we maybe include the `maven-dependency-plugin` to automatically check for broken dependencies? Here is an example including a few configured exceptions:

https://github.com/opencast/opencast/blob/3bc4af82ce74b3270b117f5bbacd93061714a728/modules/workspace-impl/pom.xml#L105-L117

This isn't strictly necessary but we have lately started to introduce this to catch dependency problems early.

lkiesow commented on modules/userdirectory-brightspace/pom.xml:
```diff
@@ -0,0 +1,153 @@
+    <packaging>bundle</packaging>
+    <name>Opencast :: userdirectory-brightspace</name>
+    <parent>
+        <groupId>org.opencastproject</groupId>
```
Not extremely important but all other `pom.xml` files use two spaces for indentation ;-)

kristofkeppens replied:
I prefer 2 spaces as well, I'll change the pom.xml :)

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+   */
+  @Override
+  public void deleted(String pid) {
+    logger.debug("delete BrightspaceUserProviderInstance for pid=" + pid);
```
```suggestion
    logger.debug("delete BrightspaceUserProviderInstance for pid={}", pid);
```

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+
+      try {
+        ManagementFactory.getPlatformMBeanServer().unregisterMBean(getObjectName(pid));
+      } catch (Exception var4) {
```
```suggestion
      } catch (Exception e) {
```

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+      try {
+        ManagementFactory.getPlatformMBeanServer().unregisterMBean(getObjectName(pid));
+      } catch (Exception var4) {
+        logger.warn("Unable to unregister mbean for pid='{}': {}", pid, var4.getMessage());
```
```suggestion
        logger.warn("Unable to unregister mbean for pid='{}'", pid, e);
```

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+      if (properties.get(CACHE_EXPIRATION) != null) {
+        return Integer.parseInt(properties.get(CACHE_EXPIRATION).toString());
+      }
+    } catch (NumberFormatException nfe) {
```
My general suggestion for configuration parsing would be to rather fail early if a configuration is broken than to fall back to a default value. That way, as a system administrator you will notice that problem immediately and can fix it instead of accidentally running with an incorrect value for whatever time ;-)

In general, I would also suggest using common-lang3's [NumberUtils::toLong(String str, long defaultValue)](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/math/NumberUtils.html#toLong-java.lang.String-long-). Makes your life much easier and basically any Opencast module uses it anyway.

lkiesow replied:
I mean [toInt(String str, int defaultValue)](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/math/NumberUtils.html#toInt-java.lang.String-int-)

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+    return DEFAULT_CACHE_EXPIRATION_VALUE;
+  }
+
+  private int parseCacheSizeProperty(Dictionary properties) {
```
same here…

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+    return DEFAULT_CACHE_SIZE_VALUE;
+  }
+
+  private void validateSystemUserKey(String userKey) throws ConfigurationException {
```
The following four methods are identical except for the key which you could just pass along to the method as arhument.

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+    if (StringUtils.isBlank(urlStr)) {
+      throw new ConfigurationException(BRIGHTSPACE_URL, "is not set");
+    } else {
+      URI url;
```
This is never used

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java:
```diff
@@ -0,0 +1,246 @@
+      URI url;
+      try {
+        url = new URI(urlStr);
+      } catch (URISyntaxException var21) {
```
```suggestion
      } catch (URISyntaxException e) {
```

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderInstance.java:
```diff
@@ -0,0 +1,274 @@
+  private String pid;
+  private BrightspaceClient client;
+  private Organization organization;
+  private LoadingCache<String, Object> cache;
```
I'm wondering if it wouldn't make sense to make this

```java
  private LoadingCache<String, User> cache;
```

Instead of casting the Objects to users later on.

kristofkeppens replied:
Seems like a valid argument :), checked the code and was copied over from the other userdirectories as well. I can try to have a look at changing it.

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderInstance.java:
```diff
@@ -0,0 +1,274 @@
+   */
+  @Override
+  public long countUsers() {
+    return 0L;
```
Wouldn't that suggest that this user provider never holds any users?

kristofkeppens replied:
I'm assuming that this was copied over from the sakai userprovider as well, not sure why this is this way.

smarquard replied:
The LMS user providers are intended to support looking up users, not "holding" users like the internal user provider does. There's no value in providing an arbitrary user count here.

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderInstance.java:
```diff
@@ -0,0 +1,274 @@
+   */
+  @Override
+  public Iterator<User> findUsers(String query, int offset, int limit) {
+    return Collections.emptyIterator();
```
Same here. Why never return any user?

kristofkeppens replied:
I think it is because we don't need this method, not sure. From where would you use this method, is this to search for users from the UI?

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/client/BrightspaceClientImpl.java:
```diff
@@ -0,0 +1,220 @@
+              });
+      return brightspaceUserList.stream().findFirst().orElse(null);
+    } catch (IOException e) {
+      logger.error(UNEXPECTED_JSON_RESPONSE);
```
No need to log this if you throw the same message as exception anyway.

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/client/BrightspaceClientImpl.java:
```diff
@@ -0,0 +1,220 @@
+      UsersResponse usersResponse = objectMapper.readValue(response, UsersResponse.class);
+      return usersResponse.getItems();
+    } catch (IOException e) {
+      logger.error(UNEXPECTED_JSON_RESPONSE);
```
same here

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/client/BrightspaceClientImpl.java:
```diff
@@ -0,0 +1,220 @@
+    } catch (MalformedURLException mue) {
+      throw new BrightspaceClientException("url was malformed", mue);
+    }
+    logger.debug("about to make GET request to : {}", uri.toString());
```
```suggestion
    logger.debug("about to make GET request to : {}", uri);
```

lkiesow commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/client/BrightspaceClientImpl.java:
```diff
@@ -0,0 +1,220 @@
+      while ((inputline = bufferedReader.readLine()) != null) {
+        content.append(inputline);
+      }
+      logger.debug("call to brightspace api: {}", content.toString());
```
```suggestion
      logger.debug("call to brightspace api: {}", content);
```

lkiesow reviewed (changes requested):
I did a quick run-through without any actual testing. Most comments aren't very important. Hard-coding `admin` is probably the most important one.

JamesUoM commented on modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/client/api/BrightspaceUser.java:
```diff
@@ -0,0 +1,103 @@
+  private String displayName;
+
+  @JsonCreator
+  public BrightspaceUser(@JsonProperty("OrgId") String orgId, @JsonProperty("UserId") String userId, @JsonProperty("FirstName") String firstName, @JsonProperty("MiddleName") String middleName, @JsonProperty("LastName") String lastName, @JsonProperty("UserName") String userName, @JsonProperty("ExternalEmail") String externalEmail, @JsonProperty("OrgDefinedId") String orgDefinedId, @JsonProperty("U ...
```
This line is too long (by about 0.5km ;-) )

kristofkeppens committed:
add dependencies to main pom and fix checkstyle erros

kristofkeppens committed:
add dependencies to main pom and fix checkstyle erros

kristofkeppens committed:
Update modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

kristofkeppens committed:
Update modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

kristofkeppens committed:
fix error message when api communication error occurs

kristofkeppens committed:
fix error message when api communication error occurs

kristofkeppens committed:
add maven-dependency-plugin and fix following warnings

kristofkeppens committed:
add maven-dependency-plugin and fix following warnings

kristofkeppens committed:
fix indentation to 2 spaces

kristofkeppens committed:
fix indentation to 2 spaces

kristofkeppens committed:
Update modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

kristofkeppens committed:
Update modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

kristofkeppens committed:
remove unnecessary logging

kristofkeppens committed:
remove unnecessary logging

kristofkeppens committed:
Change to sane variable naming

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

kristofkeppens committed:
Change to sane variable naming

Co-Authored-By: Lars Kiesow <lkiesow@uos.de>

kristofkeppens committed:
fix log messages and long strings

kristofkeppens committed:
fix log messages and long strings

kristofkeppens committed:
cleanup userproviderfactory methods

kristofkeppens committed:
cleanup userproviderfactory methods

kristofkeppens committed:
remove trailing spaces

kristofkeppens committed:
remove trailing spaces

lkiesow committed:
MH-13560, Admin Role in Moodle User Provider

The Moodle user provider assumed the username of the global admin user
to always be `admin` but the name is configurable.

lkiesow committed:
MH-13560, Central Admin User Key Definition

This patch moves the property name for the global admin user into the
SecurityConstants class instead of having it defined over and over
again.

KatrinIhler commented:
This PR has test failures!

kristofkeppens committed:
fix hardcoded admin role

kristofkeppens committed:
fix hardcoded admin role

kristofkeppens committed:
rewrite test to simplified version and remove dependency on mockito

kristofkeppens committed:
rewrite test to simplified version and remove dependency on mockito

KatrinIhler commented:
Test failures and conflicts!

gregorydlogan commented:
@flyapen Any idea when there will be movement on this?

kristofkeppens committed:
Merge branch 'develop' into feature/userdirectory-brightspace

kristofkeppens committed:
change to new log4j lib

kristofkeppens committed:
fixes change in jaxbuser constructor

kristofkeppens committed:
merge

kristofkeppens commented:
@gregorydlogan I think I've been able to resolve most comments, there are still some open issues. If these prohibit the merge we'll have a look at asking outside help with the remaining issues.

gregorydlogan commented:
I see two remaining issues, assuming no new issues were introduced:

1) The serialization library is not the same one the rest of the project uses
2) There are some missing Javadocs

I really only see 1 as a serious issue (we've got lots of missing javadocs), and I'm not sure it's worth blocking this issue.  @lkiesow, @JamesUoM, @smarquard, thoughts?

gregorydlogan commented:
Per discussion in the technical meeting today, the libraries in question are already actually in use elsewhere in the codebase, so this is going in right... now.
